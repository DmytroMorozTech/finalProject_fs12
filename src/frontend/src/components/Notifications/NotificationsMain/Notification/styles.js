import { makeStyles } from '@material-ui/core/styles'

export default makeStyles((theme) => ({

  notification: {
    display: 'flex',
    flexDirection: 'row'
  },

  userAvatar: {
    width: theme.avatar.medium,
    borderRadius: '50%',
    marginTop: theme.spacing(2),
    marginLeft: theme.spacing(3)
  },

  content: {
    margin: theme.spacing(3),
    cursor: 'pointer'
  },

  userName: {
    fontWeight: theme.typography.h5.fontWeight,
    color: theme.palette.grey[700]
  },

  link: {
    textDecoration: 'none'
  },

  actionAndText: {
    color: theme.palette.grey[700]
  },

  notificationTimeAndMenu: {
    color: theme.palette.grey[500],
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-end',
    margin: theme.spacing(2)
  },

  notificationTime: {
    fontSize: theme.typography.h6.fontSize,
    marginRight: theme.spacing(2)
  },

  btnMarkViewed: {
    marginTop: theme.spacing(3),
    color: theme.palette.primary.light,
    backgroundColor: theme.palette.secondary.contrastText,
    borderColor: theme.palette.primary.light,
    border: '1px solid',
    borderRadius: '20px'
  }
}))
