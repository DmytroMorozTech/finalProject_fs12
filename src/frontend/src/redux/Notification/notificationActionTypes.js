export const LOADING_NOTIFICATIONS = 'LOADING_NOTIFICATIONS'
export const SAVE_NEW_NOTIFICATIONS = 'SAVE_NEW_NOTIFICATIONS'
