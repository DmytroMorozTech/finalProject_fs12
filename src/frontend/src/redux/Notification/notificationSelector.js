export const notificationsSelector = (state) => state.notifications.notificationsList
