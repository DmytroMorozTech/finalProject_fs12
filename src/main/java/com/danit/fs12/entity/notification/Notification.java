package com.danit.fs12.entity.notification;

import com.danit.fs12.entity.AbstractEntity;
import com.danit.fs12.entity.user.User;
import com.vladmihalcea.hibernate.type.json.JsonStringType;
import lombok.*;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.TypeDef;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.HashMap;

@Entity
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
@Data
@Table(name = "notifications")
@TypeDef(name = "json", typeClass = JsonStringType.class)
public class Notification extends AbstractEntity {

  public Notification(
    NotificationType typeNotification,
    Long relatedUserId,
//    NotificationHashMap dataNotification,
    Long postId,
    Long likeId
  ) {
    super();
    this.typeNotification = typeNotification;
    this.relatedUserId = relatedUserId;
//    this.dataNotification = dataNotification;
    this.postId = postId;
    this.likeId = likeId;
  }

  @Column(name = "type_notification")
  private NotificationType typeNotification; // e.g. NEW_POST_WAS_CREATED  ordinal = 1

  @Type(type = "json")
  @Column(name = "data_notification", columnDefinition = "json")
  private NotificationHashMap dataNotification;

  @Column(name = "post_id")
  private Long postId;

  @Column(name = "like_id")
  private Long likeId;

  @Column(name = "related_user_id")
  private Long relatedUserId; // id of User who triggered this Notification

  @ManyToOne
  @JoinColumn(name = "user_id")
  private User user; // activeUser

}