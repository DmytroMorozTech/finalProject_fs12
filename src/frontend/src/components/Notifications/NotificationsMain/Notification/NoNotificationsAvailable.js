import Typography from '@material-ui/core/Typography'
import React from 'react'
import styles from './styles'
import Image from '../../../../shared/Image/Image'

function NoNotificationsAvailable () {
  const classes = styles()

  const noNotificationsImg = <Image
    imageUrl={'linkedin/general/mrtvwl5vxbu7hzi5m2sc'}
    alt={'no notifications'}
    className={classes.noNotificationsImg}
    type={'largeAvatar'}
  />

  return (
    <div className={classes.notification}>
      {noNotificationsImg}
      <div className={classes.content}>
        <Typography variant="body1">
          <span className={classes.headerText}>At the moment, there are no notifications for you.</span>
        </Typography>
      </div>

    </div>
  )
}

export default NoNotificationsAvailable
