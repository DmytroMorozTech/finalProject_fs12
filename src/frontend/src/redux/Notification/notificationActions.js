import http from '../../services/httpService'
import * as actions from './notificationActionTypes'
import { toast } from 'react-toastify'

export const markNotificationAsViewedAction = (payload) => (dispatch) => {
  const id = payload

  return http
    .put(`/api/notifications/mark_as_viewed/${id}`)
    .then((res) => res.data)
    .then((newNotificationObj) => {
      dispatch({ type: actions.MARK_NOTIFICATION_AS_VIEWED, payload: newNotificationObj })
    })
}

export const deleteNotificationAction = (id) => (dispatch) => {
  return http
    .delete(`/api/notifications/${id}`)
    .then((res) => res.status)
    .then((status) => {
      if (status === 200 || status === 204) {
        dispatch({ type: actions.DELETE_NOTIFICATION, payload: id })
      }
    })
}

export const markAllNotificationsAsViewedAction = () => (dispatch) => {
  return http
    .put(`/api/notifications/mark_as_viewed/all`)
    .then((res) => res.status)
    .then((status) => {
      if (status === 200 || status === 204) {
        dispatch({ type: actions.MARK_ALL_NOTIFICATIONS_AS_VIEWED })
        toast.info('You have marked all notifications as viewed',
          {position: toast.POSITION.TOP_RIGHT, autoClose: 1700})
      }
    })
}
