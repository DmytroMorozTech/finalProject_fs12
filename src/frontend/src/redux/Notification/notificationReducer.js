import * as actions from './notificationActionTypes'

const initialState = {
  notificationsList: [],
  loading: false,
  pagination: {
    pageNumber: 0,
    pageSize: 8,
    totalPages: 0,
    hasMore: true
  }
}

const notificationReducer = (state = initialState, action) => {
  switch (action.type) {
    case actions.LOADING_NOTIFICATIONS:
      return { ...state, loading: action.payload }

    case actions.SAVE_NEW_NOTIFICATIONS:
      const { notifications, headers } = action.payload
      const {pagenumber = 0, pagesize = 4, totalpages, hasmore} = headers
      return {
        ...state,
        notificationsList: [...state.notificationsList, ...notifications],
        pagination: {
          pageNumber: parseInt(pagenumber) + 1,
          pageSize: parseInt(pagesize),
          totalPages: parseInt(totalpages),
          hasMore: hasmore === 'true'
        }
      }

    default: {
      return state
    }
  }
}

export default notificationReducer
