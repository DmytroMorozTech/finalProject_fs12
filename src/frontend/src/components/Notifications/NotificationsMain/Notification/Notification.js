import styles from './styles'
import Typography from '@material-ui/core/Typography'
import SimpleMenu from '../../../../shared/PopupMenu/PopupMenu'
import ThreeDots from '../../../../shared/ThreeDots/TreeDots'
import NotificationAdditions from '../Additions/NotificationAdditions'
import React from 'react'
import { Link } from 'react-router-dom'
import Image from '../../../../shared/Image/Image'
import ChatOutlinedIcon from '@material-ui/icons/ChatOutlined'
import clsx from 'clsx'
import getTimeSinceCreated from '../../../../services/timePassedServiceShort'
import { useDispatch } from 'react-redux'
import { markNotificationAsViewedAction } from '../../../../redux/Notification/notificationActions'

function generateMainText (number, notificationType) {
  return number > 20 ? `Outstanding result! You seem to be a confident member of LinkedIn network. Number of ${notificationType}s is more than 20. `
    : number > 10 ? `Great result. Number of ${notificationType}s is more than 10.`
      : number > 5 ? `Your posts are getting more attention from community. Number of ${notificationType}s is more than 5.`
        : number > 2 ? `Great start. Number of ${notificationType}s is more than 2.`
          : 'Let the communication begin.'
}

function getHowManyTimesStr (number) {
  return number === 1 ? 'once' : number === 2 ? 'twice' : `${number} times`
}

function Notification (props) {
  const {notification} = props
  const { id, type, data, isViewed, createdDate, userWhoTriggered } = notification
  const { postId, numberOfLikes = 0, numberOfComments = 0 } = data
  const classes = styles()
  const dispatch = useDispatch()

  let headerText
  let mainText

  switch (type) {
    case 'POST_WAS_LIKED':
      headerText = `Your post with id ${postId} was liked ${getHowManyTimesStr(numberOfLikes)}`
      mainText = generateMainText(numberOfLikes, 'like')
      break

    case 'NEW_COMMENTS_POST':
      headerText = `Your post with id ${postId} was commented ${getHowManyTimesStr(numberOfComments)}`
      mainText = generateMainText(numberOfComments, 'comment')
      break

    case 'NEW_POST_WAS_CREATED':
      const {fullName, text} = userWhoTriggered
      headerText = `${fullName} has created a Post. `
      mainText = text
      break

    default:
      headerText = `You've just received a new notification`
      mainText = 'Some extra data about your notification should be here'
      break
  }

  const thumbsUpImg = <Image
    imageUrl={'linkedin/general/uj1vnnyai3jssffhgasg'}
    alt={'user avatar'}
    className={classes.notificationImg}
    type={'smallAvatar'}
  />

  const userWhoTriggeredAvatarImg = userWhoTriggered ? <Image
    imageUrl={userWhoTriggered.imgPublicId}
    alt={'user avatar'}
    className={classes.notificationImg}
    type={'mediumAvatar'}
  /> : ''

  const imgForNotification = (type === 'POST_WAS_LIKED') ? thumbsUpImg
    : (type === 'NEW_COMMENTS_POST') ? <ChatOutlinedIcon className={classes.notificationImg}/>
      : userWhoTriggeredAvatarImg

  return (
    <div key={notification.id} className={clsx(classes.notification, !isViewed ? classes.wasNotViewed : '')}>
      <div>
        {imgForNotification}
      </div>
      <div className={classes.content}>
        <Typography variant="body1">
          <Link
            to={`/posts/${postId}`}
            className={classes.link}
            onClick={() => dispatch(markNotificationAsViewedAction(id))}
          >
            <span className={classes.headerText}>{headerText}</span>
            <span className={classes.mainText}>{mainText}</span>
          </Link>
        </Typography>
      </div>

      <div className={classes.notificationTimeAndMenu}>
        <div className={classes.notificationTime}>
          {getTimeSinceCreated(createdDate)}
        </div>
        <SimpleMenu
          menuItem={<ThreeDots/>}
          userData={<NotificationAdditions notificationId={id} isViewed={isViewed}/>}/>
      </div>
    </div>
  )
}

export default Notification
