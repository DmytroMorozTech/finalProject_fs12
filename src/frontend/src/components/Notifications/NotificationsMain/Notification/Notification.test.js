import { shallow } from 'enzyme'
import React from 'react'
import Notification from './Notification'
import * as redux from 'react-redux'

jest.mock('./styles', () => () => ({}))

const setUp = (props) => shallow(<Notification {...props}/>)

describe('should render Notification component', () => {
  let spyOnUseDispatch
  let mockDispatch

  let notification

  let component
  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch')
    mockDispatch = jest.fn()
    spyOnUseDispatch.mockReturnValue(mockDispatch)

    notification = {
      'id': 20,
      'createdDate': '2021-10-13T16:26:30.95932',
      'type': 'NEW_POST_WAS_CREATED',
      'data': {
        'postId': 26
      },
      'userWhoTriggered': {
        'imgPublicId': 'linkedin/avatars/hj8sulgxxo5ywotggkcy',
        'fullName': 'Katherine McCarthy',
        'id': '5',
        'text': 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aene...'
      },
      'isViewed': false
    }

    component = setUp({ notification })
  })
  it('to match snapshot', () => {
    expect(component).toMatchSnapshot()
  })
})

describe('should render Notification component', () => {
  let spyOnUseDispatch
  let mockDispatch

  let notification

  // eslint-disable-next-line no-unused-vars
  let component
  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch')
    mockDispatch = jest.fn()
    spyOnUseDispatch.mockReturnValue(mockDispatch)

    notification = {
      'id': 13,
      'createdDate': '2021-10-05T23:10:32.123123',
      'type': 'POST_WAS_LIKED',
      'data': {
        'numberOfLikes': 19,
        'postId': 24
      },
      'userWhoTriggered': null,
      'isViewed': false
    }

    component = setUp({ notification })
  })
  it('should work case POST_WAS_LIKED', () => {
    expect.stringMatching('was liked')
  })
})

describe('should render Notification component', () => {
  let spyOnUseDispatch
  let mockDispatch

  let notification

  // eslint-disable-next-line no-unused-vars
  let component
  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch')
    mockDispatch = jest.fn()
    spyOnUseDispatch.mockReturnValue(mockDispatch)

    notification = {
      'id': 1,
      'createdDate': '2021-10-03T10:02:21.123123',
      'type': 'NEW_COMMENTS_POST',
      'data': {
        'postId': 1,
        'numberOfComments': 3
      },
      'userWhoTriggered': null,
      'isViewed': false
    }

    component = setUp({ notification })
  })
  it('should work case NEW_COMMENTS_POST', () => {
    expect.stringMatching('was commented')
  })
})

describe('should render Notification component', () => {
  let spyOnUseDispatch
  let mockDispatch

  let notification

  // eslint-disable-next-line no-unused-vars
  let component
  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch')
    mockDispatch = jest.fn()
    spyOnUseDispatch.mockReturnValue(mockDispatch)

    notification = {
      'id': 1,
      'createdDate': '2021-10-03T10:02:21.123123',
      'type': 'default',
      'data': {
        'postId': 1,
        'numberOfComments': 3
      },
      'userWhoTriggered': null,
      'isViewed': false
    }

    component = setUp({ notification })
  })
  it('should work case default', () => {
    expect.stringMatching('You\'ve just received a new notification')
  })
})
