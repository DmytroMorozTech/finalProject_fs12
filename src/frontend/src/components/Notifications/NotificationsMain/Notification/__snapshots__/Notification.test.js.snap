// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Notification component to match snapshot 1`] = `
<div
  className=""
  key="20"
>
  <div>
    <Image
      alt="user avatar"
      imageUrl="linkedin/avatars/hj8sulgxxo5ywotggkcy"
      type="mediumAvatar"
    />
  </div>
  <div>
    <WithStyles(ForwardRef(Typography))
      variant="body1"
    >
      <Link
        onClick={[Function]}
        to="/posts/26"
      >
        <span>
          Katherine McCarthy has created a Post. 
        </span>
        <span>
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aene...
        </span>
      </Link>
    </WithStyles(ForwardRef(Typography))>
  </div>
  <div>
    <div>
      6d
    </div>
    <SimpleMenu
      menuItem={<ThreeDots />}
      userData={
        <NotificationAdditions
          isViewed={false}
          notificationId={20}
        />
      }
    />
  </div>
</div>
`;
